code/modules/projectiles/guns/ballistic.dm

```diff
@@ -70,16 +70,18 @@
 			to_chat(user, "<span class='notice'>There's already a magazine in \the [src].</span>")
 	if(istype(A, /obj/item/suppressor))
 		var/obj/item/suppressor/S = A
-		if(can_suppress)
+		if(!user.is_holding(src))
```

Comment:
You should bail out on !can_supressed first, so you can avoid this check if it's not even valid